src/auth/index.ts

```
@@ -0,0 +1,144 @@
+import dotenv from 'dotenv';
+import AWS from 'aws-sdk';
+import {
+  CognitoUserPool,
+  CognitoUser,
+  AuthenticationDetails,
+} from 'amazon-cognito-identity-js';
+
+const authRouter: Router = Router();
+const parseJSON = json();
+authRouter.use(parseJSON);
+dotenv.config();
+const awsCreds = {
+  accessKeyId: process.env.ACCESS_ID,
+  SecretAccessKey: process.env.SECRET_ACCESS,
+  sessionToken: process.env.SESSION_TOKEN,
+  region: process.env.region,
+};
+
+AWS.config.update(awsCreds);
+
+authRouter.post('/login', async (req, res) => {
+  const key: Secret | undefined = process.env.ACCESS_KEY;
+  const { username, newPassword = '' } = req.body;
+
+  // Configure Cognito user pool
+  const userPool = new CognitoUserPool({
+    UserPoolId: process.env.COGNITO_USERPOOLID as string,
+    ClientId: process.env.COGNITO_CLIENTID as string,
+  });
+
+  // Create authentication details
+  const authenticationData = {
+    Username: req.body.username,
+    Password: req.body.password,
+  };
+  const authenticationDetails = new AuthenticationDetails(authenticationData);
+
+  // Create Cognito user object
+  const userData = {
+    Username: username,
+    Pool: userPool,
+  };
+
+  const cognitoUser = new CognitoUser(userData);
+
+  // Authenticate user
+
+  cognitoUser.authenticateUser(authenticationDetails, {
+    onSuccess: (session) => {
+      const options: jwt.SignOptions = {
+        header: { alg: 'HS256' },
+      };
+      const load = {
+        name: session.getIdToken().payload.email,
+        email: session.getIdToken().payload.email,
+        picture: null,
+        sub: session.getIdToken().payload.sub,
+        userId: session.getIdToken().payload.sub,
+        clientId: session.getIdToken().payload.aud,
+      };
+
+      const jwtToken = jwt.sign(load, key as string, options);
+      res.status(200).send(jwtToken);
+    },
+    onFailure: () => {
+      // Authentication failed
+      res.status(401).json({ error: 'Authentication failed' });
```

Perhaps add some additional info here. Since we are "swallowing" the original error it will be tough to actually figure out bugs later.

```typescript
onFailure: (error) => {
    // Authentication failed
    res.status(401).json({
	    error: 'Authentication failed',
	    code: error.code,
	    nestedError: error
	});
	logger.error(`Authentication failed due to ${error.message}`)
	logger.error(error)
}
```

You will have to see what fields are available in the error AWS sends back. Maybe there is a code or a message or something so we can tell the user why they couldn't sign in (connection issue, wrong credentials, etc).